feat: add roundedCorners option for BrowserWindow

zcbenz opened the pull request:
#### Description of Change

Add an option to remove rounded corners of frameless windows on macOS, close https://github.com/electron/electron/issues/414.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Add `roundedCorners` option for `BrowserWindow`.

samuelmaddock reviewed (commented):
BrowserWindows with vibrancy options seem to apply their own corner radius size. Would this new option be compatible with this behavior?
https://github.com/electron/electron/blob/95e5193c0297e728de1f532d8fde952262185e52/shell/browser/native_window_mac.mm#L1242-L1258

Would it be possible to instead provide a `cornerRadius` option where the amount could be specified? If it were `0`, rounded corners could be disabled. I'm not too familiar with macOS development to know whether this is supported or not.

miniak commented on docs/api/browser-window.md:
```diff
@@ -228,6 +228,8 @@ It creates a new `BrowserWindow` with native properties as set by the `options`.
       experimental.
   * `trafficLightPosition` [Point](structures/point.md) (optional) - Set a
     custom position for the traffic light buttons in frameless windows.
+  * `roundedCorner` Boolean (optional) - Whether frameless window should have
```
wouldn't `roundedCorners` be a better name?

zcbenz commented:
> BrowserWindows with vibrancy options seem to apply their own corner radius size. Would this new option be compatible with this behavior?

Thanks for noticing it, I have fixed the code to work with vibrancy views.

> Would it be possible to instead provide a `cornerRadius` option where the amount could be specified? If it were `0`, rounded corners could be disabled. I'm not too familiar with macOS development to know whether this is supported or not.

No, the referenced code is used to set corner radius of the vibrancy view so it would not cover the window's rounded corners, macOS does not provide a way to change the corner radius of the window.

> wouldn't roundedCorners be a better name?

Yeah it sounds a better name, I have changed the option's name.

ldenoue commented:
I noticed that when using this property `rounderCorners: true` when creating a BrowserWindow with a Tray, my app and windows never close.

giacomocerquone commented:
> I noticed that when using this property `roundedCorners: true` when creating a BrowserWindow with a Tray, my app and windows never close.

yes, a repro has been created and the issue is tracked here: https://github.com/electron/electron/issues/28686

I tried to have a look at the code, but honestly, I'm no obj-c macOS dev sadly, and couldn't figure out much.s
@zcbenz would be nice to have some hints on what to investigate upon.

Alecyrus commented:
@zcbenz  Is it possible that this option is available for windows 11?

zcbenz commented:
> @zcbenz Is it possible that this option is available for windows 11?

I can't find any Windows API for it. There is also an article on this topic: https://valinet.ro/2022/01/21/Disable-the-rounded-corners-in-Windows-11.html, but it does not imply any windows style to disable the rounded corner.

renegadevi commented:
> @zcbenz Is it possible that this option is available for windows 11?

I need a solution for this as well. Windowed mode is now "broken" in windows 11 because of rounded corners which is not always the intended use when you want a on purpose a dimensioned canvas. This issue is gone when pressing F11 to go into fullscreen mode or when using kioskmode, which is not ideal either for all use cases. At the same time, the full canvas is still recognised as a square with all the pixels even tho the pixels are like masked above.


Windows 10:

![windows_10](https://user-images.githubusercontent.com/10997469/208795789-0c83eab3-e539-45ff-94f9-94d7a1341459.png)


Windows 11: (corners being cut off)
![windows_11](https://user-images.githubusercontent.com/10997469/208795806-a484d27c-2f04-4302-9130-bf76ff36db2c.png)

renegadevi commented:
I found one workaround. By using `thickFrame: false` it fixes both the rounded corners and the position being 0,0 again. 

However, this in turn causes the F11 key only to fullscreen, I cannot longer back out from fullscreen. 

@Alecyrus @zcbenz

Pratiyankkumar commented:
I can't use roundedCorners: true in windows .(Is it just for macOS??)

wangrongding commented:
Nice! After searching for a while, I finally found it!👏👏👏
